src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionMappingService.cs

```diff
@@ -186,44 +208,38 @@ private static void AutoMapFilters(FilterMappingPlan filtersPlan)
-            {
-                var candidatesForLevel = targetLocationMeta
-                    .Any(entry => entry.Key.Level == level.Level)
-                    ? targetLocationMeta
-                        .Single(entry => entry.Key.Level == level.Level)
-                        .Value
-                    : [];
-
-                return new LocationLevelMappings
+            .ToDictionary(
+                keySelector: level => level.Level,
+                elementSelector: level =>
                 {
-                    Level = level.Level,
-                    Mappings = level
-                        .Options
-                        .OrderBy(option => option.Label)
-                        .Select(option => new LocationOptionMapping
-                        {
-                            Source = new LocationOption
-                            {
-                                Key = $"{option.Label} :: {option.ToRow().GetRowKey()}",
-                                Label = option.Label,
-                            }
-                        })
-                        .ToList(),
-                    Candidates = candidatesForLevel
-                        .OrderBy(option => option.Label)
-                        .Select(option => new LocationOption
-                        {
-                            Key = $"{option.Label} :: {option.GetRowKey()}",
-                            Label = option.Label
-                        })
-                        .ToList()
-                };
-            })
-            .ToList();
-
-        var sourceLevels = sourceMappingsByLevel.Select(level => level.Level);
+                    var candidatesForLevel = targetLocationMeta
+                        .Any(entry => entry.Key.Level == level.Level)
+                        ? targetLocationMeta
+                            .Single(entry => entry.Key.Level == level.Level)
+                            .Value
+                        : [];
+
+                    return new LocationLevelMappings
+                    {
+                        Mappings = level
+                            .Options
+                            .Select(option => option.ToRow())
+                            .ToDictionary(
+                                keySelector: LocationOptionKeyGenerator,
+                                elementSelector: option => new LocationOptionMapping
+                                {
+                                    Source = CreateLocationOptionFromMetaRow(option)
+                                }),
+                        Candidates = candidatesForLevel
+                            .ToDictionary(
+                                keySelector: LocationOptionKeyGenerator,
+                                elementSelector: CreateLocationOptionFromMetaRow)
+                    };
+                });
+
+        var sourceLevels = sourceMappingsByLevel
+            .Select(level => level.Key)
+            .Distinct();
```

Comment:
Think you can just do `var sourceLevels = sourceMappingsByLevel.Keys` here :)

Reply:
Ahh yeah, right you are again.  Changed ta